\documentclass{article}

\begin{document}


\title{Expanding our reference with sequence graphs}

\author{Erik Garrison}

\maketitle

\begin{abstract}

Despite the availability of whole genome sequences from tens of thousands of humans, when examining a new genome we only use single haplotype to drive our interpretation.
Analyses are biased against the detection of any variation comprising significant differences from this reference haplotype.
To resolve this issue, and also to expand resequencing approaches to organisms for which no single reference sequence has been (or can be) constructed, we propose the use of sequence graphs that simultaneously describe all known genomes.
\end{abstract}

\section{Introduction}

Where genomes are small and sequences from different individuals can be reliably isolated, we can understand variation by assembling whole genomes and then comparing them via whole-genome comparison approaches \cite{mummer}.
However, as the genomes of organisms of interest (such as \emph{H. sapiens}) are often large, and the genomes of organisms we want to understand are often complex and difficult to assemble reliably using existing technology, standard practice involves the localization of short reads against a single high-quality reference system that is typically composed of a single haplotype per homologous region. Effectively, this reference describes a single individual, although 


\section{Methods}

\section{Results}

\section{Discussion}

\bibliography{references}{}

\bibliographystyle{plain}



\end{document}
