\documentclass{article}
\usepackage{hyperref}
\usepackage{graphicx}
%\usepackage{algorithmicx}
\usepackage{algorithm2e}

\begin{document}

\title{Variation graphs to support DNA sequence data analysis}

\author{Erik Garrison (eg10@sanger.ac.uk)\\ supervisor: Richard Durbin (rd@sanger.ac.uk)}

\maketitle

Standard genomics approaches determine the sequence of a new sample by localizing short, cheap reads from that sample against a single high-quality reference genome of an individual from the same species. While expedient, this approach biases results towards a reference system that may poorly represent alleles present in the novel sample. We would like to align to a genome that is as similar to our sample as possible, ideally a ``personalized'' reference genome \cite{Yuan_2012}. However, there is no need for this system to be personalized to a single individual and it might be more useful if it could be used for any individual. A promising solution is to construct and then resequence against a \emph{variation graph} which contains all known genomes from our target species. In this graph we label nodes with sequences we have observed and use edges to describe allowed linkages between sequences. When constructed from known genomes, the graph will include the input genomes as walks through its elements. Alignments against this structure will not be biased toward any single genome in the ensemble from which it is built.

This project aims to develop this approach to support a more accurate model for alignment and interpretation
of genome sequence data.  A first step is to support efficient read alignment, for which a prototype exists.
This would effectively mean that most (~99\%) variant sites in a sample would be called in a standardised
fashion during alignment, speeding up calling and allowing attention to be focused on rare novel variants.
It would also support much improved calling of known/shared structural variants.  At this point we hope
that this software will become the preferred approach for analysing whole genome sequence data on the
HiSeq X platform at Sanger and elsewhere.  In the future we aim to integrate longer range genetic haplotype
structure; this would support efficient imputation-based
variant calling to get highly accurate genotypes from low coverage or uneven data.  We then would hope to
use these structure for compressed representation of very large data sets, of the order of one million human
genome sequences.  There are also interesting potential interactions with genome assembly.

A basic implementation of this model exists in a software package developed and extended during the first two PhD rotations (\url{https://github.com/ekg/vg}). This system has allowed interactive testing of operations and workflows involving variation graphs and alignments against them. However, a number of open questions remain:

\begin{enumerate}
  \item Can operations using the variation graph, such as alignment and variant calling, be done as efficiently as they are currently done when using a single, linear reference genome?
  \item Can haplotypes from a population be integrated into variant calling, genotyping, and hapolytpe determination using alignments against the graph? (Such as by using PBWT \cite{durbin2014pbwt}.)
  \item Can the variation graph be used for RNA sequencing analysis?
  \item What are the optimal models for the exchange of information and relating to the graph, such as sequence annotations (e.g. of functional elements or genes), haplotypes (whole genomes), and genotypes?
  \item How can the graph be extended as our perspective on the genome of a particular species improves?
  \item Could such a system be used to model uncertain sequences, such as might be important in the process of structural variant detection or genotyping?
\end{enumerate}

To explore solutions to these questions and drive the work towards a production-quality result, we will apply the following strategy:

\begin{enumerate}
  \item Develop a succinct model for the graph (such as using structures from \url{https://github.com/simongog/sdsl-lite}) to improve memory usage and performance when querying and traversing the graph.
  \item Utilize GCSA for sequence queries of the graph \cite{sireÃÅn2014gcsa}.
  \item Support simultaneous phasing (haplotyping) and genotyping by developing or using a queryable haplotype reference (such as PBWT).
  \item Develop methods to extend the graph using alignments, either short read or contig/BAC/unitig alignments such as much be used when developing a reference system for high-diversity regions such as the MHC.
  \item Test alignment against graphs generated by assembly methods such as Fermikit \cite{li2015fermikit}.
  \item Develop graphs for RNA sequencing experiments.
\end{enumerate}

\bibliography{references}{}

\bibliographystyle{plain}




\end{document}
